[github-actions] buildx-codelab_otsim build failing

Build started breaking after merging #6454 

```
#9 63.62 ../../src/posix/platform/libopenthread-posix.a(libopenthread_posix_a-system.o): In function `otSysInit':
#9 63.62 /root/src/openthread/src/posix/../../src/posix/platform/system.cpp:116: undefined reference to `__cxa_guard_acquire'
#9 63.62 /root/src/openthread/src/posix/../../src/posix/platform/system.cpp:116: undefined reference to `__cxa_guard_release'
#9 63.62 collect2: error: ld returned 1 exit status
```

[logs_77027.zip](https://github.com/openthread/openthread/files/6481235/logs_77027.zip)

--- a/.github/workflows/docker.yml
+++ b/.github/workflows/docker.yml
@@ -66,6 +66,7 @@ jobs:
         echo ::set-output name=docker_image::${DOCKER_IMAGE}
         echo ::set-output name=version::${VERSION}
         echo ::set-output name=buildx_args::--platform ${DOCKER_PLATFORMS} \
+          --build-arg OT_GIT_REF=${{ github.sha }} \
           --build-arg VERSION=${VERSION} \
           --build-arg BUILD_DATE=$(date -u +'%Y-%m-%dT%H:%M:%SZ') \
           --build-arg VCS_REF=${GITHUB_SHA::8} \

--- a/etc/docker/codelab_otsim/Dockerfile
+++ b/etc/docker/codelab_otsim/Dockerfile
@@ -2,6 +2,7 @@
 FROM ubuntu:18.04
 
 ENV DEBIAN_FRONTEND noninteractive
+ARG OT_GIT_REF=main
 
 # Install dependencies:
 RUN apt-get update -qq
@@ -12,10 +13,11 @@ RUN apt-get --no-install-recommends install -y git sudo software-properties-comm
     && update-ca-certificates
 
 # Install OpenThread
-RUN mkdir -p ~/src && \
-    cd ~/src && \
-    git clone --recursive https://github.com/openthread/openthread.git && \
-    cd openthread && \
+RUN mkdir -p ~/src/openthread && \
+    cd ~/src/openthread && \
+    git init . && \
+    git fetch https://github.com/openthread/openthread.git ${OT_GIT_REF} && \
+    git checkout FETCH_HEAD && \
     ./script/bootstrap && \
     ./bootstrap && \
     make -f examples/Makefile-simulation

--- a/src/posix/platform/platform-posix.h
+++ b/src/posix/platform/platform-posix.h
@@ -160,10 +160,10 @@ void platformAlarmAdvanceNow(uint64_t aDelta);
  * @note Even when @p aPlatformConfig->mResetRadio is false, a reset event (i.e. a PROP_LAST_STATUS between
  * [SPINEL_STATUS_RESET__BEGIN, SPINEL_STATUS_RESET__END]) is still expected from RCP.
  *
- * @param[in]  aPlatformConfig  Platform configuration structure.
+ * @param[in]   aUrl  A pointer to the null-terminated radio URL.
  *
  */
-void platformRadioInit(otUrl *aRadioUrl);
+void platformRadioInit(const char *aUrl);
 
 /**
  * This function shuts down the radio service used by OpenThread.

--- a/src/posix/platform/radio.cpp
+++ b/src/posix/platform/radio.cpp
@@ -35,7 +35,9 @@
 
 #include <string.h>
 
+#include "common/new.hpp"
 #include "lib/spinel/radio_spinel.hpp"
+#include "posix/platform/radio.hpp"
 
 #if OPENTHREAD_POSIX_CONFIG_RCP_BUS == OT_POSIX_RCP_BUS_UART
 #include "hdlc_interface.hpp"
@@ -53,59 +55,54 @@ static ot::Spinel::RadioSpinel<ot::Posix::SpiInterface, RadioProcessContext> sRa
 #error "OPENTHREAD_POSIX_CONFIG_RCP_BUS only allows OT_POSIX_RCP_BUS_UART and OT_POSIX_RCP_BUS_SPI!"
 #endif
 
-void otPlatRadioGetIeeeEui64(otInstance *aInstance, uint8_t *aIeeeEui64)
+namespace ot {
+namespace Posix {
+
+namespace {
+alignas(alignof(ot::Posix::Radio)) char sRadioRaw[sizeof(ot::Posix::Radio)];
+
+extern "C" void platformRadioInit(const char *aUrl)
 {
-    OT_UNUSED_VARIABLE(aInstance);
-    SuccessOrDie(sRadioSpinel.GetIeeeEui64(aIeeeEui64));
+    Radio &radio = *(new (&sRadioRaw) Radio(aUrl));
+
+    radio.Init();
+}
+} // namespace
+
+Radio::Radio(const char *aUrl)
+    : mRadioUrl(aUrl)
+{
+    VerifyOrDie(mRadioUrl.GetPath() != nullptr, OT_EXIT_INVALID_ARGUMENTS);
 }
 
-void otPlatRadioSetPanId(otInstance *aInstance, uint16_t panid)
+void Radio::Init(void)
 {
-    OT_UNUSED_VARIABLE(aInstance);
-    SuccessOrDie(sRadioSpinel.SetPanId(panid));
-}
-
-void otPlatRadioSetExtendedAddress(otInstance *aInstance, const otExtAddress *aAddress)
-{
-    OT_UNUSED_VARIABLE(aInstance);
-    otExtAddress addr;
-
-    for (size_t i = 0; i < sizeof(addr); i++)
-    {
-        addr.m8[i] = aAddress->m8[sizeof(addr) - 1 - i];
-    }
-
-    SuccessOrDie(sRadioSpinel.SetExtendedAddress(addr));
-}
-
-void otPlatRadioSetShortAddress(otInstance *aInstance, uint16_t aAddress)
-{
-    OT_UNUSED_VARIABLE(aInstance);
-    SuccessOrDie(sRadioSpinel.SetShortAddress(aAddress));
-}
-
-void otPlatRadioSetPromiscuous(otInstance *aInstance, bool aEnable)
-{
-    OT_UNUSED_VARIABLE(aInstance);
-    SuccessOrDie(sRadioSpinel.SetPromiscuous(aEnable));
-}
-
-void platformRadioInit(otUrl *aRadioUrl)
-{
-    ot::Url::Url &radioUrl               = *static_cast<ot::Url::Url *>(aRadioUrl);
-    bool          resetRadio             = (radioUrl.GetValue("no-reset") == nullptr);
-    bool          restoreDataset         = (radioUrl.GetValue("ncp-dataset") != nullptr);
-    bool          skipCompatibilityCheck = (radioUrl.GetValue("skip-rcp-compatibility-check") != nullptr);
-    const char *  parameterValue;
-    const char *  region;
+    bool        resetRadio             = (mRadioUrl.GetValue("no-reset") == nullptr);
+    bool        restoreDataset         = (mRadioUrl.GetValue("ncp-dataset") != nullptr);
+    bool        skipCompatibilityCheck = (mRadioUrl.GetValue("skip-rcp-compatibility-check") != nullptr);
+    const char *parameterValue;
+    const char *region;
 #if OPENTHREAD_POSIX_CONFIG_MAX_POWER_TABLE_ENABLE
     const char *maxPowerTable;
 #endif
 
-    SuccessOrDie(sRadioSpinel.GetSpinelInterface().Init(radioUrl));
+#if OPENTHREAD_POSIX_VIRTUAL_TIME
+    // The last argument must be the node id
+    {
+        const char *nodeId = nullptr;
+
+        for (const char *arg = nullptr; (arg = mRadioUrl.GetValue("forkpty-arg", arg)) != nullptr; nodeId = arg)
+        {
+        }
+
+        virtualTimeInit(static_cast<uint16_t>(atoi(nodeId)));
+    }
+#endif
+
+    SuccessOrDie(sRadioSpinel.GetSpinelInterface().Init(mRadioUrl));
     sRadioSpinel.Init(resetRadio, restoreDataset, skipCompatibilityCheck);
 
-    parameterValue = radioUrl.GetValue("fem-lnagain");
+    parameterValue = mRadioUrl.GetValue("fem-lnagain");
     if (parameterValue != nullptr)
     {
         long femLnaGain = strtol(parameterValue, nullptr, 0);
@@ -114,7 +111,7 @@ void platformRadioInit(otUrl *aRadioUrl)
         SuccessOrDie(sRadioSpinel.SetFemLnaGain(static_cast<int8_t>(femLnaGain)));
     }
 
-    parameterValue = radioUrl.GetValue("cca-threshold");
+    parameterValue = mRadioUrl.GetValue("cca-threshold");
     if (parameterValue != nullptr)
     {
         long ccaThreshold = strtol(parameterValue, nullptr, 0);
@@ -123,7 +120,7 @@ void platformRadioInit(otUrl *aRadioUrl)
         SuccessOrDie(sRadioSpinel.SetCcaEnergyDetectThreshold(static_cast<int8_t>(ccaThreshold)));
     }
 
-    region = radioUrl.GetValue("region");
+    region = mRadioUrl.GetValue("region");
     if (region != nullptr)
     {
         uint16_t regionCode;
@@ -134,7 +131,7 @@ void platformRadioInit(otUrl *aRadioUrl)
     }
 
 #if OPENTHREAD_POSIX_CONFIG_MAX_POWER_TABLE_ENABLE
-    maxPowerTable = radioUrl.GetValue("max-power-table");
+    maxPowerTable = mRadioUrl.GetValue("max-power-table");
     if (maxPowerTable != nullptr)
     {
         constexpr int8_t kPowerDefault = 30; // Default power 1 watt (30 dBm).
@@ -171,9 +168,49 @@ void platformRadioInit(otUrl *aRadioUrl)
 #endif // OPENTHREAD_POSIX_CONFIG_MAX_POWER_TABLE_ENABLE
 }
 
+} // namespace Posix
+} // namespace ot
+
 void platformRadioDeinit(void)
 {
     sRadioSpinel.Deinit();
+}
+
+void otPlatRadioGetIeeeEui64(otInstance *aInstance, uint8_t *aIeeeEui64)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    SuccessOrDie(sRadioSpinel.GetIeeeEui64(aIeeeEui64));
+}
+
+void otPlatRadioSetPanId(otInstance *aInstance, uint16_t panid)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    SuccessOrDie(sRadioSpinel.SetPanId(panid));
+}
+
+void otPlatRadioSetExtendedAddress(otInstance *aInstance, const otExtAddress *aAddress)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    otExtAddress addr;
+
+    for (size_t i = 0; i < sizeof(addr); i++)
+    {
+        addr.m8[i] = aAddress->m8[sizeof(addr) - 1 - i];
+    }
+
+    SuccessOrDie(sRadioSpinel.SetExtendedAddress(addr));
+}
+
+void otPlatRadioSetShortAddress(otInstance *aInstance, uint16_t aAddress)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    SuccessOrDie(sRadioSpinel.SetShortAddress(aAddress));
+}
+
+void otPlatRadioSetPromiscuous(otInstance *aInstance, bool aEnable)
+{
+    OT_UNUSED_VARIABLE(aInstance);
+    SuccessOrDie(sRadioSpinel.SetPromiscuous(aEnable));
 }
 
 bool otPlatRadioIsEnabled(otInstance *aInstance)

--- a/src/posix/platform/radio.hpp
+++ b/src/posix/platform/radio.hpp
@@ -1,0 +1,65 @@
+/*
+ *  Copyright (c) 2021, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#ifndef POSIX_PLATFORM_RADIO_HPP_
+#define POSIX_PLATFORM_RADIO_HPP_
+
+#include "posix/platform/radio_url.hpp"
+
+namespace ot {
+namespace Posix {
+
+/**
+ * This class manages Thread radio.
+ *
+ */
+class Radio
+{
+public:
+    /**
+     * This method creates the radio manager.
+     *
+     * @param[in]   aUrl    A pointer to the null-terminated URL.
+     *
+     */
+    explicit Radio(const char *aUrl);
+
+    /**
+     * This method initialize the Thread radio.
+     *
+     */
+    void Init(void);
+
+private:
+    RadioUrl mRadioUrl;
+};
+
+} // namespace Posix
+} // namespace ot
+
+#endif // POSIX_PLATFORM_RADIO_HPP_

--- a/src/posix/platform/system.cpp
+++ b/src/posix/platform/system.cpp
@@ -111,26 +111,9 @@ static const char *getTrelRadioUrl(otPlatformConfig *aPlatformConfig)
 otInstance *otSysInit(otPlatformConfig *aPlatformConfig)
 {
     otInstance *instance = nullptr;
-    // radioUrl must be static to prevent from having dangling pointers after this function call ends
-    // and we still need to refer to the radioUrl from HdlcInterface object
-    static ot::Posix::RadioUrl radioUrl(get802154RadioUrl(aPlatformConfig));
 
-#if OPENTHREAD_POSIX_VIRTUAL_TIME
-    // The last argument must be the node id
-    {
-        const char *nodeId = nullptr;
-
-        for (const char *arg = nullptr; (arg = radioUrl.GetValue("forkpty-arg", arg)) != nullptr; nodeId = arg)
-        {
-        }
-
-        virtualTimeInit(static_cast<uint16_t>(atoi(nodeId)));
-    }
-#endif
-
-    VerifyOrDie(radioUrl.GetPath() != nullptr, OT_EXIT_INVALID_ARGUMENTS);
     platformAlarmInit(aPlatformConfig->mSpeedUpFactor, aPlatformConfig->mRealTimeSignal);
-    platformRadioInit(&radioUrl);
+    platformRadioInit(get802154RadioUrl(aPlatformConfig));
 #if OPENTHREAD_CONFIG_RADIO_LINK_TREL_ENABLE
     platformTrelInit(getTrelRadioUrl(aPlatformConfig));
 #endif

